frontend/packages/pipelines-plugin/src/components/pipelines/detail-page-tabs/pipeline-details/PipelineVisualizationTask.tsx

```diff
@@ -261,8 +263,8 @@ const TaskComponent: React.FC<TaskProps> = ({
         <WhenExpressionDecorator
           width={WHEN_EXPRESSSION_DIAMOND_SIZE}
           height={WHEN_EXPRESSSION_DIAMOND_SIZE}
-          stroke={showStatusState ? taskColor : undefined}
-          color={taskColor}
+          stroke={showStatusState ? whenExpressionDiamondColor : undefined}
+          color={whenExpressionDiamondColor}
```

Comment:
This is not added by you, but I don't understand why we make the color calculation outside and the tooltip text decision in `WhenExpressionDecorator`.

I'm not sure, but I think it's possible to remove stroke and color props here and calculate these in the `switch (status)` the component already have. You maybe need a new prop for `isFinallyTask={isFinallyTask}`

@debsmita1 @karthikjeeyar @andrewballantyne Wdyt? Is there an edge case I don't see yet? Or is there any reason to calculate these data on different component levels?

Reply:
There are two variables  that decide the color/stroke/backgroundColor of the whenExpressionDecorator in P/PL details page.
1. showStatusState - is used to figure out if it is in pipeline details page or pipeline run details page.
2. isFinallyTask - is used to have a different background color between regular/finally task.

Either these two variables has to be passed down the component instead of color and stroke and calculate it on WhenExpressionDecorator or the caller of the WhenExpressionDecorator can decide the color/stroke and the decorator doesn't need to know if its finally task or regular task and about pipeline/pipelineRun page. I went with latter when I first implemented when expressions.

Reply:
Okay when it is possible I **personally** would recommend encapsulate this calculation. It is done also 20 lines above for the `taskPill`.

```tsx
            <PipelineVisualizationStepList
              isSpecOverview={!isPipelineRun}
              taskName={visualName}
              steps={stepStatusList}
              isFinallyTask={isFinallyTask}
            />
```

From my perspective we don't need a `showStatusState` prop. The `status` prop is already optional and a Task in a Pipeline doesn't have a status, right? So `status` and `isFinallyTask` should be enough.

My hope is also that a switch-case block or static object is much easier to read that the multiple ternary operators here:

```tsx
  const whenExpressionDiamondColor = showStatusState
    ? status.reason === runStatus.Failed
      ? getRunStatusColor(runStatus.Succeeded).pftoken.value
      : taskStatusColor
    : !isFinallyTask
    ? greyBackgroundColor.value
    : lightBackgroundColor.value;
```

I also recommend to import the colors in `WhenExpressionDecorator` then directly from PF instead of reusing `getRunStatusColor`. The condition `reason === Failed ? SuccessColor : taskStatusColor` really confused me at the first moment.

And btw: `enableTooltip` prop is always true.

Everything is just an opinion. It's up to you both, Debsmita and Karthik. But could not add approve this without at least mentioned this. :smirk:

Reply:
I agree that this is could be improved.